rfc/hierarchical-capabilities.md

```diff
@@ -0,0 +1,57 @@
+# Hierarchical Capabilities
+
+## Abstract
+
+Currently UCAN capabilities described in [W3Up specification](https://github.com/web3-storage/specs) do not follow consistent organizational structure. To make things worse accounts on login delegate `*` capabilities and spaces delegate set of namespaces.
+
+This creates a serious challenge for the programs (like w3up client) that attempt to infer set of methods to expose based no delegated capabilities.
+
+We roughly two roles in the sysntem `space` and `account` and their capability set do not overlap, so we can somewhat infer which methods to present in case of `*` ability by subject DID, but that hinges upon the fact that we use `did:key` for space and `did:mailto` for accounts. This assumbtion may be rendered obsolete in the future.
+
+## Proposal
+
+Re-organize capabilities into role based namespeces e.g. `/space/` and `/account/` making it possible for the program to infer which set of methods to expose. All of the `store/*` and `upload/*` etc capabilities SHOULD be remapped under `/space/` namespace.
+
+In addition we should capture depndency across namespaces in the way capabilities are orginalizated e.g. if upload task requires storing bytes we nest `store` capabilities under `upload` here is an example layout that may work
+
+
+```sh
+# write permission
+/space/content/add                          # upload/add
+/space/content/add/shard                    # store/add
+/space/content/add/shard/allocate           # store/allocate
+/space/content/add/shard/write              # S3/R2 PUT
+
+# read permission
+/space/content/list                         # upload/list
+/space/content/list/shard                   # store/list
+
+# delete permission
+/space/content/remove                       # upload/remove
+/space/content/remove/shard                 # store/remove
+
+# write+delete permission
+/space/content/edit/add                     # upload/add
+/space/content/edit/add/shard               # store/add
+/space/content/edit/add/shard/allocate      # store/allocate
```

Comment:
Can you please elaborate what exactly does not ? Are you suggesting that no one would want to give `write + delete` or that calling it `edit` is misleading or something else ?

Reply:
The fact that there are 3 different verbs in 1 capability - edit, add and allocate - I do not understand what being delegated this capability would allow me to do.